docs/proposals/virtualhost-policy.md

```diff
@@ -0,0 +1,98 @@
+# VirtualHost policy
+
+## Context
+
+Since Kuma 1.1.4 the dataplane now embeds a CoreDNS instance which forwards all DNS traffic to envoy proxy.
+This opens the door to more flexible implementation of vips.
+We propose here to create a new policy which will enable users to create arbitrary hostnames and ports that will route to a specific sets of tags.
+
+## New Policy
+
+We will add a new policy type:
+
+```yaml
+type: VirtualHost
+mesh: string # The mesh this policy applies to
+name: string #The name of the policy
+tags:
+    # A map of tags that should be taken for creating the tag selector. 
+    # It can either have a value or have "*" in which case a different vip will be generated for each tuple
+host: string # A go template with a special function `tag` which can use tags present in the `tags` section.
+port: int # A port to expose the virtual host on
+```
+
+It is impossible to create a VirtualHost that hasn't got a tag `kuma,io/service`.
+
+### Example
+
+Simple service with no wildcard replacement:
+
+```yaml
+type: VirtualHost
+mesh: default
+name: my-service
+tags:
+  kuma.io/service: my-service
+host: httpbin.mesh
+port: 8080
+```
+
+Create different hostnames for each version of a service:
+
+```yaml
+type: VirtualHost
+mesh: default
+name: my-service-by-version
+tags:
+  kuma.io/service: "*"
+  version: "*"
+host: "{{tag version}}.{{tag kuma.io/service}}.mesh"
+port: 8080
+```
+
+A policy to enable things to be backward compatible:
+
+```yaml
+type: VirtualHost
+mesh: default
+name: my-service-by-version
+tags:
+  kuma.io/service: "*"
+host: "{{tag kuma.io/service}}.mesh"
+port: 80
+```
+
+A virtual host to get per zone domains
+
+```yaml
+type: VirtualHost
+mesh: default
+name: my-service-by-zone
+tags:
+  kuma.io/service: "*"
+  kuma.io/zone: "*"
+host: "{{tag kuma.io/service}}.{{kuma.io/zone}}.mesh"
+port: 80
+```
+
+
```

Comment:
What if I do this

```
type: VirtualHost
mesh: default
name: 1
tags:
  kuma.io/service: "*"
host: "{{tag kuma.io/service}}.{{tag version}}.mesh"
port: 80
```
and there is no version? Will this host for service be omitted?

If so, we could support multizone access to individual instance of the application
```
type: VirtualHost
mesh: default
name: 1
tags:
  kuma.io/service: "*"
host: "{{tag kuma.io/service}}.{{tag kuma.io/instance}}.mesh"
port: 80
```
something that is missing today

Reply:
I was planning on having only keys present in the tags section.
If we don't do this what's the point of the tags section?
Should it instead be a `selector` to restrict which dps get a virtualHost?

If was planning to use the tags tuple for indexing but maybe <generatedHost>:<port>:<virtualHostName> is enough?
I need to think about this some more. Let me know if you have some answers to these questions.

Reply:
After some thinking.
`tags` will become `selector: match` similar to TracingPolicy.
I think it's reasonable for all keys used in the template to have to be present in the selector.